### PR TITLE
Waypoint system

### PR DESCRIPTION
**Pull request by Mpcs**
# Description
 Waypoint system.
 They render in world and as a border marker. One can remove and add waypoints using buttons on the map screen. Waypoints are saved with hero.
 Also added UI sliders.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: engine/src/main/java/org/destinationsol/game/Hero.java [hunk 1]
```diff
@@ -154,6 +162,20 @@ public Vector2 getPosition() {
         return isTranscendent ? transcendentHero.getPosition() : shipHero.getPosition();
     }
 
+    public void addWaypoint(Waypoint waypoint) {
+        waypoints.add(waypoint);
+        System.out.println("ADDED WAYPOINT");
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/Hero.java]: Is this intended to be here?

File: engine/src/main/java/org/destinationsol/game/MapDrawer.java [hunk 2]
```diff
@@ -134,7 +139,7 @@ public void draw(GameDrawer drawer, SolGame game) {
         drawPlanets(drawer, game, viewDist, np, camPos, heroDmgCap, camAngle);
         drawMazes(drawer, game, viewDist, np, camPos, heroDmgCap, camAngle);
         drawStarNodes(drawer, game, viewDist, camPos, starNodeW);
-
+        drawWaypoints(drawer, game, iconSz, viewDist);
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/MapDrawer.java]: Maybe add the newline back here after the `drawWaypoints` call?

File: engine/src/main/java/org/destinationsol/game/SaveManager.java [hunk 3]
```diff
@@ -72,6 +95,7 @@ public static void writeShips(HullConfig hull, float money, List<SolItem> itemsL
      * @param items A list of SolItems to be encoded as a string
      * @return A string of items suitable for saving
      */
+
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/SaveManager.java]: Why is there a newline here?

File: engine/src/main/java/org/destinationsol/game/SaveManager.java [hunk 4]
```diff
@@ -180,7 +204,6 @@ public static boolean hasPrevShip(String fileName) {
      */
     public static ShipConfig readShip(HullConfigManager hullConfigs, ItemManager itemManager) {
         IniReader ir = new IniReader(SAVE_FILE_NAME, null);
-
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/SaveManager.java]: Why was this line deleted?

File: engine/src/main/java/org/destinationsol/game/screens/BorderDrawer.java [hunk 5]
```diff
@@ -121,6 +122,43 @@ public void draw(UiDrawer drawer, SolApplication application) {
...
+            float size;
+            if(closeness < 0.5f) {
+                size = BORDER_ICON_SZ * 0.5f;
+            } else {
+                size = BORDER_ICON_SZ * closeness;
+            }
+
+            Vector2 position = new Vector2();
+            SolMath.toRel(waypoint.position, position, cam.getAngle(), camPos);
+
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/screens/BorderDrawer.java]: Any reason for two newlines here?

File: engine/src/main/java/org/destinationsol/game/screens/BorderDrawer.java [hunk 6]
```diff
@@ -149,7 +187,6 @@ private void maybeDrawIcon(UiDrawer drawer, Vector2 position, SolCam cam, float
         float mul = SolMath.abs(prefXAxis ? (prefX / myTmpVec.x) : (prefY / myTmpVec.y));
         myTmpVec.scl(mul);
         myTmpVec.add(displayDimensions.getRatio() / 2, .5f);
-
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/screens/BorderDrawer.java]: why was this line removed?

File: engine/src/main/java/org/destinationsol/game/screens/MapScreen.java [hunk 7]
```diff
@@ -76,5 +101,94 @@ public void updateCustom(SolApplication solApplication, SolInputManager.InputPoi
...
+        if (removeWaypointControl.isJustOff() && !isPickingWaypointToRemove) {
+            isPickingWaypointToRemove = true;
+            removeWaypointControl.setDisplayName(CANCEL_TEXT);
+            addWaypointControl.setEnabled(false);
+        } else if (removeWaypointControl.isJustOff() && isPickingWaypointToRemove) {
+            removeWaypointControl.setDisplayName(REMOVE_WAYPOINT_TEXT);
+            addWaypointControl.setEnabled(true);
+            isPickingWaypointToRemove = false;
+        }
+        else if (isPickingWaypointToRemove) {
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/screens/MapScreen.java]: The formatting seems a bit inconsistent here. Shouldn't the elseif be one line up?

File: engine/src/main/java/org/destinationsol/game/screens/MapScreen.java [hunk 8]
```diff
@@ -76,5 +101,94 @@ public void updateCustom(SolApplication solApplication, SolInputManager.InputPoi
...
+
+        Vector2 finalPosition = new Vector2(camPos);
+        finalPosition.add(screenPos);
+
+        finalPosition.x -= (ratio * mapZoom)/(2.f/5);
+        finalPosition.y -= (mapZoom)/(2.f/5);
+        return finalPosition;
+    }
+
+    public void disableWaypointButtons(boolean value) {
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/game/screens/MapScreen.java]: This function can be simplied by removing the if-Statement and then changing the calls to `setEnabled(!value)`.
(If there is a setDisabled() function, use it instead.)

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    private final Rectangle rectangle;
+    private float value;
+    private String text;
+    private final int trimAt;
+
+    public SolUiSlider(Rectangle sliderRectangle, String text, float startingValue, int trimAt) {
+        rectangle = sliderRectangle;
+        this.text = text;
+        if (startingValue > 1.0f) {
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: I think using Math.min()/Math.max() could provide more readibility here and in the following lines.

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    public void draw(UiDrawer uiDrawer) {
+        uiDrawer.draw(rectangle, SolColor.UI_INACTIVE);
+        uiDrawer.draw(sliderTexture, rectangle.width, rectangle.height, 0, 0, rectangle.x, rectangle.y, 0, SolColor.WHITE);
+
+        uiDrawer.setTextMode(true);
+        String trimmedValue = Float.toString(value);
+        int lenght = trimmedValue.substring(trimmedValue.indexOf('.')).length();
+        System.out.println(value);
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: Is it intended to print this here?

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            clickPosition.y > rectangle.y && clickPosition.y < rectangle.y + rectangle.height) {
+            float relativePos = clickPosition.x - rectangle.x;
+            setValue((relativePos/rectangle.width));
+            return true;
+        }
+        return false;
+    }
+
+    public void setValue(float val) {
+        if(val < 0f) {
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: Dito about Math.min()/Math.max().

File: engine/src/main/java/org/destinationsol/ui/Waypoint.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            public boolean shouldBeRemoved(SolGame game) {
+                return false;
+            }
+
+            @Override
+            public SolObject toObject(SolGame game) {
+                for(Waypoint waypoint : game.getHero().getWaypoints())
+                    if(waypoint.position.x == position.x && waypoint.position.y == position.y)
+                        return waypoint;
+                    System.out.println("RETURNDED NUL L");
```
Comment (FearlessTobi) [inline engine/src/main/java/org/destinationsol/ui/Waypoint.java]: I doubt this was intended to be here.

File: engine/src/main/java/org/destinationsol/game/Hero.java (hunk 1 above)
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/Hero.java]: If yes, please, at least `Log` instead

File: engine/src/main/java/org/destinationsol/game/screens/MapScreen.java [hunk 13]
```diff
@@ -76,5 +101,94 @@ public void updateCustom(SolApplication solApplication, SolInputManager.InputPoi
...
+            if (inputPointers[0].isJustUnPressed()) {
+                Vector2 clickPosition = new Vector2(inputPointers[0].x, inputPointers[0].y);
+
+                Vector2 worldPosition = screenPositionToWorld(clickPosition, game.getCam().getPosition(), mapZoom);
+
+                ArrayList<Waypoint> waypoints = game.getHero().getWaypoints();
+                boolean canCreate = true;
+                for (int w = 0; w < waypoints.size(); w++) {
+                    Waypoint waypoint = waypoints.get(w);
+                    if (worldPosition.x > waypoint.position.x-MIN_WAYPOINT_DIST && worldPosition.x < waypoint.position.x+MIN_WAYPOINT_DIST &&
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/screens/MapScreen.java]:
```suggestion
                    if (worldPosition.x > waypoint.position.x - MIN_WAYPOINT_DIST && worldPosition.x < waypoint.position.x + MIN_WAYPOINT_DIST &&
```

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java (hunk 11 above)
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: `SolMath.clamp()`

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java (hunk 10 above)
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: `Log`

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        this.trimAt = trimAt;
+    }
+
+    public void draw(UiDrawer uiDrawer) {
+        uiDrawer.draw(rectangle, SolColor.UI_INACTIVE);
+        uiDrawer.draw(sliderTexture, rectangle.width, rectangle.height, 0, 0, rectangle.x, rectangle.y, 0, SolColor.WHITE);
+
+        uiDrawer.setTextMode(true);
+        String trimmedValue = Float.toString(value);
+        int lenght = trimmedValue.substring(trimmedValue.indexOf('.')).length();
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: `length` (a common typo)

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java (hunk 9 above)
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]: `SolMath.clamp()`

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        uiDrawer.drawString(text + trimmedValue, rectangle.x + rectangle.width * 0.5f, rectangle.y + rectangle.height * 0.5f, 1, true, SolColor.WHITE);
+        uiDrawer.setTextMode(false);
+        uiDrawer.draw(sliderMarkerTexture, rectangle.width * HANDLE_SCALE, rectangle.height, rectangle.width * HANDLE_SCALE * 0.5f, 0, rectangle.x + value * rectangle.width, rectangle.y, 0, SolColor.WHITE);
+    }
+
+    public boolean click(Vector2 clickPosition) {
+        if(clickPosition.x > rectangle.x && clickPosition.x < rectangle.x + rectangle.width &&
+            clickPosition.y > rectangle.y && clickPosition.y < rectangle.y + rectangle.height) {
+            float relativePos = clickPosition.x - rectangle.x;
+            setValue((relativePos/rectangle.width));
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]:
```suggestion
            setValue(relativePos / rectangle.width);
```

File: engine/src/main/java/org/destinationsol/ui/Waypoint.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.destinationsol.ui;
+
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.g2d.TextureAtlas;
+import com.badlogic.gdx.math.Vector2;
+import org.destinationsol.game.*;
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/ui/Waypoint.java]: Do not use star imports for anything (disable them in your IDE for our projects at least)

File: engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+package org.destinationsol.game.screens;
+
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.math.Rectangle;
+import com.badlogic.gdx.math.Vector2;
+import org.destinationsol.GameOptions;
+import org.destinationsol.SolApplication;
+import org.destinationsol.common.SolColor;
+import org.destinationsol.menu.MenuLayout;
+import org.destinationsol.ui.*;
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java]: Star import

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by Mpcs** [comment]
Prints weren't supposed to be in the code. I missed them

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/destinationsol/game/PlayerCreator.java [hunk 18]
```diff
@@ -70,6 +73,29 @@ private void addAndEquipItems(Hero hero, RespawnState respawnState, SolGame game
         itemContainer.markAllAsSeen();
     }
 
+    private void addWaypoints(Hero hero, String waypoints, RespawnState respawnState, SolGame game) {
+        if(waypoints == "") {
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/game/PlayerCreator.java]:
```suggestion
        if (waypoints == "") {
```

File: engine/src/main/java/org/destinationsol/game/screens/MapScreen.java [hunk 19]
```diff
@@ -76,5 +101,88 @@ public void updateCustom(SolApplication solApplication, SolInputManager.InputPoi
...
+                    }
+                }
+                isPickingWaypointToRemove = false;
+                addWaypointControl.setEnabled(true);
+            }
+        }
+    }
+
+    public Vector2 screenPositionToWorld(Vector2 screenPos, Vector2 camPos, float mapZoom) {
+        float ratio = (float)Gdx.graphics.getWidth() / (float)Gdx.graphics.getHeight();
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/game/screens/MapScreen.java]:
```suggestion
        float ratio = (float) Gdx.graphics.getWidth() / (float) Gdx.graphics.getHeight();
```

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public void draw(UiDrawer uiDrawer) {
+        uiDrawer.draw(rectangle, SolColor.UI_INACTIVE);
+        uiDrawer.draw(sliderTexture, rectangle.width, rectangle.height, 0, 0, rectangle.x, rectangle.y, 0, SolColor.WHITE);
+
+        uiDrawer.setTextMode(true);
+        String trimmedValue = Float.toString(value);
+        int length = trimmedValue.substring(trimmedValue.indexOf('.')).length();
+        if(length > trimAt) {
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]:
```suggestion
        if (length > trimAt) {
```

File: engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    public void drawImages(UiDrawer uiDrawer, SolApplication solApplication) {
+        uiDrawer.draw(previewRect, outcomeColor);
+        sliderRed.draw(uiDrawer);
+        sliderGreen.draw(uiDrawer);
+        sliderBlue.draw(uiDrawer);
+    }
+
+    @Override
+    public void updateCustom(SolApplication solApplication, SolInputManager.InputPointer[] inputPointers, boolean clickedOutside) {
+        if(doneControl.isJustOff()) {
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java]:
```suggestion
        if (doneControl.isJustOff()) {
```

File: engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+            Waypoint waypoint = new Waypoint(waypointPos, outcomeColor, solApplication.getGame().getMapDrawer().getWaypointTexture());
+            solApplication.getGame().getHero().addWaypoint(waypoint);
+            solApplication.getGame().getObjectManager().addObjDelayed(waypoint);
+            solApplication.getInputManager().setScreen(solApplication, solApplication.getGame().getScreens().mapScreen);
+            mapScreen.setWaypointButtonsEnabled(true);
+        }
+        
+        if(cancelControl.isJustOff()) {
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java]:
```suggestion
        if (cancelControl.isJustOff()) {
```

File: engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+            solApplication.getInputManager().setScreen(solApplication, solApplication.getGame().getScreens().mapScreen);
+            mapScreen.setWaypointButtonsEnabled(true);
+        }
+
+        if(inputPointers[0].pressed) {
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/game/screens/WaypointCreationScreen.java]:
```suggestion
        if (inputPointers[0].pressed) {
```

File: engine/src/main/java/org/destinationsol/ui/SolUiSlider.java [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+            trimmedValue = leftSubstring + rightSubstring;
+        }
+
+        uiDrawer.drawString(text + trimmedValue, rectangle.x + rectangle.width * 0.5f, rectangle.y + rectangle.height * 0.5f, 1, true, SolColor.WHITE);
+        uiDrawer.setTextMode(false);
+        uiDrawer.draw(sliderMarkerTexture, rectangle.width * HANDLE_SCALE, rectangle.height, rectangle.width * HANDLE_SCALE * 0.5f, 0, rectangle.x + value * rectangle.width, rectangle.y, 0, SolColor.WHITE);
+    }
+
+    public boolean click(Vector2 clickPosition) {
+        if(clickPosition.x > rectangle.x && clickPosition.x < rectangle.x + rectangle.width &&
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/ui/SolUiSlider.java]:
```suggestion
        if (clickPosition.x > rectangle.x && clickPosition.x < rectangle.x + rectangle.width &&
```

File: engine/src/main/java/org/destinationsol/ui/Waypoint.java [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+        FarObject farObject = new FarObject() {
+            @Override
+            public boolean shouldBeRemoved(SolGame game) {
+                return false;
+            }
+
+            @Override
+            public SolObject toObject(SolGame game) {
+                for (Waypoint waypoint : game.getHero().getWaypoints())
+                    if (waypoint.position.x == position.x && waypoint.position.y == position.y) {
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/ui/Waypoint.java]: Can we just do `waypoint.position.equals(position)`? Honestly not sure but I expect position is just a Vector2

It's better practice in general to use `.equals()` due to the way `==` does the comparison.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/destinationsol/game/MapDrawer.java [hunk 26]
```diff
@@ -409,6 +423,20 @@ public void drawObjIcon(float iconSz, Vector2 position, float objAngle,
         }
     }
 
+    public void drawWaypointIcon(float iconSz, Vector2 position, TextureAtlas.AtlasRegion icon, Object drawer, Color color) {
+        float innerIconSz = iconSz * INNER_ICON_PERC;
+        //allow usage of UiDrawer or GameDrawer
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/game/MapDrawer.java]: This is an example of a comment that describes *what* the code is doing. It's a good start, but by just reading the code I can figure that out. A good comment describes not what but *why* the code is doing what it's doing. IE why could `drawer` be more than one type of object, and why is handling them the same way (after the cast) fine?

Comment (Mpcs) [reply engine/src/main/java/org/destinationsol/game/MapDrawer.java]: So should I write something like "drawer can be either UiDrawer or GameDrawer because the function is also called from BorderDrawer which has UiDrawer, not GameDrawer available"
Or just "[...] because it is called from different contexts"?

Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/MapDrawer.java]: I would lean towards something like the first option for this specific case

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by NicholasBatesNZ** [comment]
Well I would have loved to have merged this today but it does not seem to work for me...
For a start, when I click the 'Marker+` button, it does turn to cancel and 'Marker-' greys out, however nothing else happens Oo. I discovered that if I now click the greyed out 'Marker-', the waypoint creation dialog appears, and seems to work. Except that it does not place the waypoint where I am... it places it miles off to right out in the void. I went and found the closest one that I managed to place and the drawings etc. all worked well. I could not get 'Marker-' to work at all. :/

**Comment by Mpcs** [comment]
There's now text when creating or removing a waypoint. Also cleaned the control code a bit

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!
